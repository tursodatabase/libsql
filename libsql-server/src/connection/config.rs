use parking_lot::Mutex;
use serde::{Deserialize, Serialize};
use std::path::{Path, PathBuf};
use std::sync::Arc;
use std::{fs, io};
use url::Url;

use crate::error::Error;
use crate::{Result, LIBSQL_PAGE_SIZE};

#[derive(Debug)]
pub struct DatabaseConfigStore {
    config_path: PathBuf,
    tmp_config_path: PathBuf,
    config: Mutex<Arc<DatabaseConfig>>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DatabaseConfig {
    #[serde(default)]
    pub block_reads: bool,
    #[serde(default)]
    pub block_writes: bool,
    /// The reason why operations are blocked. This will be included in [`Error::Blocked`].
    #[serde(default)]
    pub block_reason: Option<String>,
    /// maximum db size (in pages)
    #[serde(default = "default_max_size")]
    pub max_db_pages: u64,
    #[serde(default)]
    pub heartbeat_url: Option<Url>,
    #[serde(default)]
    pub bottomless_db_id: Option<String>,
}

const fn default_max_size() -> u64 {
    bytesize::ByteSize::pb(1000).as_u64() / LIBSQL_PAGE_SIZE
}

impl Default for DatabaseConfig {
    fn default() -> Self {
        Self {
            block_reads: Default::default(),
            block_writes: Default::default(),
            block_reason: Default::default(),
            max_db_pages: default_max_size(),
            heartbeat_url: None,
            bottomless_db_id: None,
        }
    }
}

impl DatabaseConfigStore {
    pub fn load(db_path: &Path) -> Result<Self> {
        let config_path = db_path.join("config.json");
        let tmp_config_path = db_path.join("config.json~");

        let config = match fs::read(&config_path) {
            Ok(data) => serde_json::from_slice(&data)?,
            Err(err) if err.kind() == io::ErrorKind::NotFound => DatabaseConfig::default(),
            Err(err) => return Err(Error::IOError(err)),
        };

        Ok(Self {
            config_path,
            tmp_config_path,
            config: Mutex::new(Arc::new(config)),
        })
    }

    #[cfg(test)]
    pub fn new_test() -> Self {
        Self {
            config_path: "".into(),
            tmp_config_path: "".into(),
            config: Mutex::new(Arc::new(DatabaseConfig::default())),
        }
    }

    pub fn get(&self) -> Arc<DatabaseConfig> {
        self.config.lock().clone()
    }

    pub fn store(&self, config: DatabaseConfig) -> Result<()> {
        let data = serde_json::to_vec_pretty(&config)?;
        fs::write(&self.tmp_config_path, data)?;
        fs::rename(&self.tmp_config_path, &self.config_path)?;
        *self.config.lock() = Arc::new(config);
        Ok(())
    }
}
