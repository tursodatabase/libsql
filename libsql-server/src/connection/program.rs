use std::sync::Arc;
use std::time::{Duration, Instant};

use metrics::{histogram, increment_counter};
use rusqlite::StatementStatus;

use crate::auth::Permission;
use crate::error::Error;
use crate::metrics::{READ_QUERY_COUNT, WRITE_QUERY_COUNT};
use crate::namespace::{NamespaceName, ResolveNamespacePathFn};
use crate::query::Query;
use crate::query_analysis::StmtKind;
use crate::query_result_builder::QueryResultBuilder;

use super::config::DatabaseConfig;
use super::RequestContext;

#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
pub struct Program {
    pub steps: Arc<Vec<Step>>,
}

impl Program {
    pub fn new(steps: Vec<Step>) -> Self {
        Self {
            steps: Arc::new(steps),
        }
    }

    pub fn is_read_only(&self) -> bool {
        self.steps.iter().all(|s| s.query.stmt.is_read_only())
    }

    pub fn steps(&self) -> &[Step] {
        self.steps.as_slice()
    }

    #[cfg(test)]
    pub fn seq(stmts: &[&str]) -> Self {
        use crate::{query::Params, query_analysis::Statement};

        let mut steps = Vec::with_capacity(stmts.len());
        for stmt in stmts {
            let step = Step {
                cond: None,
                query: Query {
                    stmt: Statement::parse(stmt).next().unwrap().unwrap(),
                    params: Params::empty(),
                    want_rows: true,
                },
            };

            steps.push(step);
        }

        Self::new(steps)
    }
}

#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
pub struct Step {
    pub cond: Option<Cond>,
    pub query: Query,
}

#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
pub enum Cond {
    Ok { step: usize },
    Err { step: usize },
    Not { cond: Box<Self> },
    Or { conds: Vec<Self> },
    And { conds: Vec<Self> },
    IsAutocommit,
}

#[derive(Debug, Clone)]
pub struct DescribeResponse {
    pub params: Vec<DescribeParam>,
    pub cols: Vec<DescribeCol>,
    pub is_explain: bool,
    pub is_readonly: bool,
}

#[derive(Debug, Clone)]
pub struct DescribeParam {
    pub name: Option<String>,
}

#[derive(Debug, Clone)]
pub struct DescribeCol {
    pub name: String,
    pub decltype: Option<String>,
}

pub struct Vm<'a, B, F, S> {
    results: Vec<bool>,
    builder: B,
    program: &'a Program,
    current_step: usize,
    should_block: F,
    update_stats: S,
    resolve_attach_path: ResolveNamespacePathFn,
}

impl<'a, B, F, S> Vm<'a, B, F, S>
where
    B: QueryResultBuilder,
    F: Fn(&StmtKind) -> (bool, Option<String>),
    S: Fn(String, u64, u64, u64, Duration),
{
    pub fn new(
        builder: B,
        program: &'a Program,
        should_block: F,
        update_stats: S,
        resolve_attach_path: ResolveNamespacePathFn,
    ) -> Self {
        Self {
            results: Vec::with_capacity(program.steps().len()),
            builder,
            program,
            current_step: 0,
            should_block,
            update_stats,
            resolve_attach_path,
        }
    }

    #[inline]
    fn current_step(&self) -> &Step {
        &self.program.steps()[self.current_step]
    }

    pub fn step(&mut self, conn: &rusqlite::Connection) -> crate::Result<bool> {
        match self.try_step(conn) {
            Ok(res) => {
                self.results.push(res);
                self.current_step += 1;
                Ok(self.current_step < self.program.steps().len())
            }
            Err(e) => Err(e),
        }
    }

    fn try_step(&mut self, conn: &rusqlite::Connection) -> crate::Result<bool> {
        self.builder.begin_step()?;
        let mut enabled = match self.current_step().cond.as_ref() {
            Some(cond) => match eval_cond(cond, &self.results, conn.is_autocommit()) {
                Ok(enabled) => enabled,
                Err(e) => {
                    self.builder.step_error(e).unwrap();
                    false
                }
            },
            None => true,
        };

        let (affected_row_count, last_insert_rowid) = if enabled {
            match self.execute_query(conn) {
                // builder error interrupt the execution of query. we should exit immediately.
                Err(e @ Error::BuilderError(_)) => return Err(e),
                Err(mut e) => {
                    if let Error::RusqliteError(err) = e {
                        let extended_code =
                            unsafe { rusqlite::ffi::sqlite3_extended_errcode(conn.handle()) };

                        e = Error::RusqliteErrorExtended(err, extended_code as i32);
                    };

                    self.builder.step_error(e)?;
                    enabled = false;
                    (0, None)
                }
                Ok(x) => x,
            }
        } else {
            (0, None)
        };

        self.builder
            .finish_step(affected_row_count, last_insert_rowid)?;

        Ok(enabled)
    }

    fn prepare_attach_query(&self, attached: &str, attached_alias: &str) -> crate::Result<String> {
        let attached = attached.strip_prefix('"').unwrap_or(attached);
        let attached = attached.strip_suffix('"').unwrap_or(attached);
        let attached = NamespaceName::from_string(attached.into())?;
        let path = (self.resolve_attach_path)(&attached)?;
        let query = format!(
            "ATTACH DATABASE 'file:{}?mode=ro' AS \"{attached_alias}\"",
            path.join("data").display()
        );
        tracing::trace!("ATTACH rewritten to: {query}");
        Ok(query)
    }

    fn execute_query(&mut self, conn: &rusqlite::Connection) -> crate::Result<(u64, Option<i64>)> {
        tracing::debug!("executing query: {}", self.current_step().query.stmt.stmt);

        increment_counter!("libsql_server_libsql_query_execute");

        let start = Instant::now();
        let (blocked, reason) = (self.should_block)(&self.current_step().query.stmt.kind);
        if blocked {
            return Err(Error::Blocked(reason));
        }

        let mut stmt = if matches!(self.current_step().query.stmt.kind, StmtKind::Attach(_)) {
            match &self.current_step().query.stmt.attach_info {
                Some((attached, attached_alias)) => {
                    // nope nope nope: only builder error should return
                    let query = self.prepare_attach_query(attached, attached_alias)?;
                    conn.prepare(&query)?
                }
                None => {
                    return Err(Error::Internal(format!(
                        "Failed to ATTACH: {:?}",
                        self.current_step().query.stmt.attach_info
                    )))
                }
            }
        } else {
            conn.prepare(&self.current_step().query.stmt.stmt)?
        };

        if stmt.readonly() {
            READ_QUERY_COUNT.increment(1);
        } else {
            WRITE_QUERY_COUNT.increment(1);
        }

        let cols = stmt.columns();
        let cols_count = cols.len();
        self.builder.cols_description(cols.iter())?;
        drop(cols);

        self.current_step()
            .query
            .params
            .bind(&mut stmt)
            .map_err(Error::LibSqlInvalidQueryParams)?;

        let mut qresult = stmt.raw_query();

        let mut values_total_bytes = 0;
        self.builder.begin_rows()?;
        while let Some(row) = qresult.next()? {
            self.builder.begin_row()?;
            for i in 0..cols_count {
                let val = row.get_ref(i)?;
                values_total_bytes += value_size(&val);
                self.builder.add_row_value(val)?;
            }
            self.builder.finish_row()?;
        }
        histogram!("libsql_server_returned_bytes", values_total_bytes as f64);

        self.builder.finish_rows()?;

        // sqlite3_changes() is only modified for INSERT, UPDATE or DELETE; it is not reset for SELECT,
        // but we want to return 0 in that case.
        let affected_row_count = match self.current_step().query.stmt.is_iud {
            true => conn.changes(),
            false => 0,
        };

        // sqlite3_last_insert_rowid() only makes sense for INSERTs into a rowid table. we can't detect
        // a rowid table, but at least we can detect an INSERT
        let last_insert_rowid = match self.current_step().query.stmt.is_insert {
            true => Some(conn.last_insert_rowid()),
            false => None,
        };

        drop(qresult);

        let query_duration = start.elapsed();

        let rows_read = stmt.get_status(StatementStatus::RowsRead) as u64;
        let rows_written = stmt.get_status(StatementStatus::RowsWritten) as u64;
        let mem_used = stmt.get_status(StatementStatus::MemUsed) as u64;

        (self.update_stats)(
            self.current_step().query.stmt.stmt.clone(),
            rows_read,
            rows_written,
            mem_used,
            query_duration,
        );

        self.builder
            .add_stats(rows_read, rows_written, query_duration);

        Ok((affected_row_count, last_insert_rowid))
    }

    pub fn builder(&mut self) -> &mut B {
        &mut self.builder
    }

    /// advance the program without executing the step
    pub(crate) fn advance(&mut self) {
        self.current_step += 1;
    }

    pub(crate) fn finished(&self) -> bool {
        self.current_step >= self.program.steps().len()
    }

    pub(crate) fn into_builder(self) -> B {
        self.builder
    }
}

fn eval_cond(cond: &Cond, results: &[bool], is_autocommit: bool) -> crate::Result<bool> {
    let get_step_res = |step: usize| -> crate::Result<bool> {
        let res = results.get(step).ok_or(Error::InvalidBatchStep(step))?;
        Ok(*res)
    };

    Ok(match cond {
        Cond::Ok { step } => get_step_res(*step)?,
        Cond::Err { step } => !get_step_res(*step)?,
        Cond::Not { cond } => !eval_cond(cond, results, is_autocommit)?,
        Cond::And { conds } => conds.iter().try_fold(true, |x, cond| {
            eval_cond(cond, results, is_autocommit).map(|y| x & y)
        })?,
        Cond::Or { conds } => conds.iter().try_fold(false, |x, cond| {
            eval_cond(cond, results, is_autocommit).map(|y| x | y)
        })?,
        Cond::IsAutocommit => is_autocommit,
    })
}

fn value_size(val: &rusqlite::types::ValueRef) -> usize {
    use rusqlite::types::ValueRef;
    match val {
        ValueRef::Null => 0,
        ValueRef::Integer(_) => 8,
        ValueRef::Real(_) => 8,
        ValueRef::Text(s) => s.len(),
        ValueRef::Blob(b) => b.len(),
    }
}

pub fn check_program_auth(
    ctx: &RequestContext,
    pgm: &Program,
    config: &DatabaseConfig,
) -> crate::Result<()> {
    for step in pgm.steps() {
        match &step.query.stmt.kind {
            StmtKind::TxnBegin
            | StmtKind::TxnEnd
            | StmtKind::Read
            | StmtKind::Savepoint
            | StmtKind::Release => {
                ctx.auth.has_right(&ctx.namespace, Permission::Read)?;
            }
            StmtKind::DDL if config.shared_schema_name.is_some() => {
                ctx.auth().ddl_permitted(&ctx.namespace)?;
            }
            StmtKind::DDL | StmtKind::Write => {
                ctx.auth().has_right(&ctx.namespace, Permission::Write)?;
            }
            StmtKind::Attach(ref ns) => {
                ctx.auth.has_right(ns, Permission::AttachRead)?;
                if !ctx.meta_store.handle(ns.clone()).get().allow_attach {
                    return Err(Error::NotAuthorized(format!(
                        "Namespace `{ns}` doesn't allow attach"
                    )));
                }
            }
            StmtKind::Detach => (),
        }
    }

    Ok(())
}

pub fn check_describe_auth(ctx: RequestContext) -> crate::Result<()> {
    ctx.auth().has_right(ctx.namespace(), Permission::Read)?;
    Ok(())
}
