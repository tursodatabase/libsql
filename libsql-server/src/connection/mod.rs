use libsql_replication::rpc::replication::NAMESPACE_METADATA_KEY;
use std::ops::Deref;
use std::sync::atomic::{AtomicUsize, Ordering};
use std::sync::Arc;
use tokio::time::{Duration, Instant};

use futures::Future;
use tokio::{sync::Semaphore, time::timeout};
use tonic::metadata::BinaryMetadataValue;

use crate::auth::Authenticated;
use crate::error::Error;
use crate::metrics::{
    CONCURRENT_CONNECTIONS_COUNT, CONNECTION_ALIVE_DURATION, CONNECTION_CREATE_TIME,
};
use crate::namespace::meta_store::MetaStore;
use crate::namespace::NamespaceName;
use crate::query::{Params, Query};
use crate::query_analysis::Statement;
use crate::query_result_builder::{IgnoreResult, QueryResultBuilder};
use crate::replication::FrameNo;
use crate::Result;

use self::program::{Cond, DescribeResponse, Program, Step};

pub mod config;
pub mod connection_manager;
pub mod dump;
pub mod libsql;
pub mod program;
pub mod write_proxy;

#[cfg(not(test))]
const TXN_TIMEOUT: Duration = Duration::from_secs(5);
#[cfg(test)]
const TXN_TIMEOUT: Duration = Duration::from_millis(100);

#[derive(Clone)]
pub struct RequestContext {
    /// Authentication for this request
    auth: Authenticated,
    /// current namespace
    namespace: NamespaceName,
    meta_store: MetaStore,
}

impl RequestContext {
    pub fn new(auth: Authenticated, namespace: NamespaceName, meta_store: MetaStore) -> Self {
        Self {
            auth,
            namespace,
            meta_store,
        }
    }

    pub fn upgrade_grpc_request<T>(&self, req: &mut tonic::Request<T>) {
        let namespace = BinaryMetadataValue::from_bytes(self.namespace.as_slice());
        req.metadata_mut()
            .insert_bin(NAMESPACE_METADATA_KEY, namespace);
        self.auth.upgrade_grpc_request(req);
    }

    pub fn namespace(&self) -> &NamespaceName {
        &self.namespace
    }

    pub fn auth(&self) -> &Authenticated {
        &self.auth
    }
}

#[async_trait::async_trait]
pub trait Connection: Send + Sync + 'static {
    /// Executes a query program
    async fn execute_program<B: QueryResultBuilder>(
        &self,
        pgm: Program,
        ctx: RequestContext,
        response_builder: B,
        replication_index: Option<FrameNo>,
    ) -> Result<B>;

    /// Execute all the queries in the batch sequentially.
    /// If an query in the batch fails, the remaining queries are ignores, and the batch current
    /// transaction (if any) is rolledback.
    async fn execute_batch_or_rollback<B: QueryResultBuilder>(
        &self,
        batch: Vec<Query>,
        ctx: RequestContext,
        result_builder: B,
        replication_index: Option<FrameNo>,
    ) -> Result<B> {
        let batch_len = batch.len();
        let mut steps = make_batch_program(batch);

        if !steps.is_empty() {
            // We add a conditional rollback step if the last step was not successful.
            steps.push(Step {
                query: Query {
                    stmt: Statement::parse("ROLLBACK").next().unwrap().unwrap(),
                    params: Params::empty(),
                    want_rows: false,
                },
                cond: Some(Cond::Not {
                    cond: Box::new(Cond::Ok {
                        step: steps.len() - 1,
                    }),
                }),
            })
        }

        let pgm = Program::new(steps);

        // ignore the rollback result
        let builder = result_builder.take(batch_len);
        let builder = self
            .execute_program(pgm, ctx, builder, replication_index)
            .await?;

        Ok(builder.into_inner())
    }

    /// Execute all the queries in the batch sequentially.
    /// If an query in the batch fails, the remaining queries are ignored
    async fn execute_batch<B: QueryResultBuilder>(
        &self,
        batch: Vec<Query>,
        ctx: RequestContext,
        result_builder: B,
        replication_index: Option<FrameNo>,
    ) -> Result<B> {
        let steps = make_batch_program(batch);
        let pgm = Program::new(steps);
        self.execute_program(pgm, ctx, result_builder, replication_index)
            .await
    }

    async fn rollback(&self, ctx: RequestContext) -> Result<()> {
        self.execute_batch(
            vec![Query {
                stmt: Statement::parse("ROLLBACK").next().unwrap().unwrap(),
                params: Params::empty(),
                want_rows: false,
            }],
            ctx,
            IgnoreResult,
            None,
        )
        .await?;

        Ok(())
    }

    /// Parse the SQL statement and return information about it.
    async fn describe(
        &self,
        sql: String,
        ctx: RequestContext,
        replication_index: Option<FrameNo>,
    ) -> Result<Result<DescribeResponse>>;

    /// Check whether the connection is in autocommit mode.
    async fn is_autocommit(&self) -> Result<bool>;

    /// Calls for database checkpoint (if supported).
    async fn checkpoint(&self) -> Result<()>;

    // Calls for database vacuum (if supported).
    async fn vacuum_if_needed(&self) -> Result<()>;

    fn diagnostics(&self) -> String;
}

fn make_batch_program(batch: Vec<Query>) -> Vec<Step> {
    let mut steps = Vec::with_capacity(batch.len());
    for (i, query) in batch.into_iter().enumerate() {
        let cond = if i > 0 {
            // only execute if the previous step was a success
            Some(Cond::Ok { step: i - 1 })
        } else {
            None
        };

        let step = Step { cond, query };
        steps.push(step);
    }
    steps
}

#[async_trait::async_trait]
pub trait MakeConnection: Send + Sync + 'static {
    type Connection: Connection;

    /// Create a new connection of type Self::Connection
    async fn create(&self) -> Result<Self::Connection, Error>;

    fn throttled(
        self,
        semaphore: Arc<Semaphore>,
        timeout: Option<Duration>,
        max_total_response_size: u64,
        max_concurrent_requests: u64,
    ) -> MakeThrottledConnection<Self>
    where
        Self: Sized,
    {
        MakeThrottledConnection::new(
            semaphore,
            self,
            timeout,
            max_total_response_size,
            max_concurrent_requests,
        )
    }

    fn map<F, T>(self, f: F) -> Map<Self, F>
    where
        F: Fn(Self::Connection) -> T + Send + Sync + 'static,
        Self: Sized,
    {
        Map { inner: self, f }
    }
}

pub struct Map<T, F> {
    inner: T,
    f: F,
}

#[async_trait::async_trait]
impl<F, T, O> MakeConnection for Map<T, F>
where
    F: Fn(T::Connection) -> O + Send + Sync + 'static,
    T: MakeConnection,
    O: Connection,
{
    type Connection = O;

    async fn create(&self) -> Result<Self::Connection, Error> {
        let conn = self.inner.create().await?;
        Ok((self.f)(conn))
    }
}

#[async_trait::async_trait]
impl<T: MakeConnection> MakeConnection for Arc<T> {
    type Connection = T::Connection;

    async fn create(&self) -> Result<Self::Connection, Error> {
        self.as_ref().create().await
    }
}

#[async_trait::async_trait]
impl<F, C, Fut> MakeConnection for F
where
    F: Fn() -> Fut + Send + Sync + 'static,
    Fut: Future<Output = Result<C, Error>> + Send,
    C: Connection + Sync + Send + 'static,
{
    type Connection = C;

    async fn create(&self) -> Result<Self::Connection, Error> {
        let db = (self)().await?;
        Ok(db)
    }
}

pub struct MakeThrottledConnection<F> {
    semaphore: Arc<Semaphore>,
    connection_maker: F,
    timeout: Option<Duration>,
    // Max memory available for responses. High memory pressure
    // will result in reducing concurrency to prevent out-of-memory errors.
    max_total_response_size: u64,
    waiters: AtomicUsize,
    max_concurrent_requests: u64,
}

impl<F> MakeThrottledConnection<F> {
    fn new(
        semaphore: Arc<Semaphore>,
        connection_maker: F,
        timeout: Option<Duration>,
        max_total_response_size: u64,
        max_concurrent_requests: u64,
    ) -> Self {
        Self {
            semaphore,
            connection_maker,
            timeout,
            max_total_response_size,
            waiters: AtomicUsize::new(0),
            max_concurrent_requests,
        }
    }

    // How many units should be acquired from the semaphore,
    // depending on current memory pressure.
    fn units_to_take(&self) -> u32 {
        let total_response_size = crate::query_result_builder::TOTAL_RESPONSE_SIZE
            .load(std::sync::atomic::Ordering::Relaxed) as u64;
        if total_response_size * 2 > self.max_total_response_size {
            tracing::trace!("High memory pressure, reducing concurrency");
            16
        } else if total_response_size * 4 > self.max_total_response_size {
            tracing::trace!("Medium memory pressure, reducing concurrency");
            4
        } else {
            1
        }
    }
}

struct WaitersGuard<'a> {
    pub waiters: &'a AtomicUsize,
}

impl<'a> WaitersGuard<'a> {
    fn new(waiters: &'a AtomicUsize) -> Self {
        waiters.fetch_add(1, Ordering::Relaxed);
        Self { waiters }
    }
}

impl Drop for WaitersGuard<'_> {
    fn drop(&mut self) {
        self.waiters.fetch_sub(1, Ordering::Relaxed);
    }
}

#[async_trait::async_trait]
impl<F: MakeConnection> MakeConnection for MakeThrottledConnection<F> {
    type Connection = TrackedConnection<F::Connection>;

    async fn create(&self) -> Result<Self::Connection, Error> {
        let before_create = Instant::now();
        // If the memory pressure is high, request more units to reduce concurrency.
        tracing::trace!(
            "Available semaphore units: {}",
            self.semaphore.available_permits()
        );
        let units = self.units_to_take();
        let waiters_guard = WaitersGuard::new(&self.waiters);
        if (waiters_guard.waiters.load(Ordering::Relaxed) as u64) >= self.max_concurrent_requests {
            return Err(Error::TooManyRequests);
        }
        let fut = self.semaphore.clone().acquire_many_owned(units);
        let mut permit = match self.timeout {
            Some(t) => timeout(t, fut).await.map_err(|_| Error::DbCreateTimeout)?,
            None => fut.await,
        }
        .expect("semaphore closed");

        let units = self.units_to_take();
        if units > 1 {
            tracing::debug!("Reacquiring {units} units due to high memory pressure");
            let fut = self.semaphore.clone().acquire_many_owned(units);
            let mem_permit = match self.timeout {
                Some(t) => timeout(t, fut).await.map_err(|_| Error::DbCreateTimeout)?,
                None => fut.await,
            }
            .expect("semaphore closed");
            permit.merge(mem_permit);
        }

        let inner = self.connection_maker.create().await?;

        CONCURRENT_CONNECTIONS_COUNT.increment(1.0);
        CONNECTION_CREATE_TIME.record(before_create.elapsed());

        Ok(TrackedConnection {
            permit,
            inner,
            created_at: Instant::now(),
        })
    }
}

#[derive(Debug)]
pub struct TrackedConnection<DB> {
    inner: DB,
    #[allow(dead_code)] // just hold on to it
    permit: tokio::sync::OwnedSemaphorePermit,
    created_at: Instant,
}

impl<T> Drop for TrackedConnection<T> {
    fn drop(&mut self) {
        CONCURRENT_CONNECTIONS_COUNT.decrement(1.0);
        CONNECTION_ALIVE_DURATION.record(self.created_at.elapsed());
    }
}

impl<T> Deref for TrackedConnection<T> {
    type Target = T;

    fn deref(&self) -> &Self::Target {
        &self.inner
    }
}

#[async_trait::async_trait]
impl<DB: Connection> Connection for TrackedConnection<DB> {
    #[inline]
    async fn execute_program<B: QueryResultBuilder>(
        &self,
        pgm: Program,
        ctx: RequestContext,
        builder: B,
        replication_index: Option<FrameNo>,
    ) -> crate::Result<B> {
        self.inner
            .execute_program(pgm, ctx, builder, replication_index)
            .await
    }

    #[inline]
    async fn describe(
        &self,
        sql: String,
        ctx: RequestContext,
        replication_index: Option<FrameNo>,
    ) -> crate::Result<crate::Result<DescribeResponse>> {
        self.inner.describe(sql, ctx, replication_index).await
    }

    #[inline]
    async fn is_autocommit(&self) -> crate::Result<bool> {
        self.inner.is_autocommit().await
    }

    #[inline]
    async fn checkpoint(&self) -> Result<()> {
        self.inner.checkpoint().await
    }

    #[inline]
    async fn vacuum_if_needed(&self) -> Result<()> {
        self.inner.vacuum_if_needed().await
    }

    #[inline]
    fn diagnostics(&self) -> String {
        self.inner.diagnostics()
    }
}

#[cfg(test)]
pub mod test {
    use super::*;

    #[derive(Debug)]
    struct DummyDb;

    #[async_trait::async_trait]
    impl Connection for DummyDb {
        async fn execute_program<B: QueryResultBuilder>(
            &self,
            _pgm: Program,
            _ctx: RequestContext,
            _builder: B,
            _replication_index: Option<FrameNo>,
        ) -> crate::Result<B> {
            unreachable!()
        }

        async fn describe(
            &self,
            _sql: String,
            _ctx: RequestContext,
            _replication_index: Option<FrameNo>,
        ) -> crate::Result<crate::Result<DescribeResponse>> {
            unreachable!()
        }

        async fn is_autocommit(&self) -> crate::Result<bool> {
            unreachable!()
        }

        async fn checkpoint(&self) -> Result<()> {
            unreachable!()
        }

        async fn vacuum_if_needed(&self) -> Result<()> {
            unreachable!()
        }

        fn diagnostics(&self) -> String {
            "dummy".into()
        }
    }

    #[tokio::test]
    async fn throttle_db_creation() {
        let factory = (|| async { Ok(DummyDb) }).throttled(
            Arc::new(Semaphore::new(10)),
            Some(Duration::from_millis(100)),
            u64::MAX,
            u64::MAX,
        );

        let mut conns = Vec::with_capacity(10);
        for _ in 0..10 {
            conns.push(factory.create().await.unwrap())
        }

        assert!(factory.create().await.is_err());

        drop(conns);

        assert!(factory.create().await.is_ok());
    }
}
