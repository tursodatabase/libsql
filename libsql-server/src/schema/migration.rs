use rusqlite::Savepoint;

use crate::connection::program::{Program, Vm};
use crate::namespace::meta_store::MigrationTask;
use crate::query_result_builder::QueryResultBuilder;

use super::{Error, MigrationTaskStatus};

pub fn setup_migration_table(conn: &mut rusqlite::Connection) -> Result<(), Error> {
    conn.execute(
        "CREATE TABLE IF NOT EXISTS __libsql_migration_tasks (
            job_id INTEGER PRIMARY KEY,
            status INTEGER,
            migration TEXT NOT NULL,
            error TEXT
        )",
        (),
    )?;

    Ok(())
}

pub fn enqueue_migration_task(
    conn: &rusqlite::Connection,
    task: &MigrationTask,
    migration: &Program,
) -> Result<(), Error> {
    let migration = serde_json::to_string(migration).unwrap();
    conn.execute(
        "INSERT INTO __libsql_migration_tasks (job_id, status, migration) VALUES (?, ?, ?)",
        (task.job_id(), *task.status() as u64, &migration),
    )?;
    Ok(())
}

pub fn step_migration_task_run(
    conn: &rusqlite::Connection,
    task: &MigrationTask,
) -> Result<(), Error> {
    conn.execute(
        "
        UPDATE __libsql_migration_tasks
        SET status = ?
        WHERE job_id = ? AND status = ?
        ",
        (
            task.job_id(),
            MigrationTaskStatus::Run as u64,
            MigrationTaskStatus::DryRunSuccess as u64,
        ),
    )?;

    Ok(())
}

pub fn perform_migration<B: QueryResultBuilder>(
    conn: &mut rusqlite::Transaction,
    migration: &Program,
    dry_run: bool,
    builder: B,
) -> (Result<B, Error>, MigrationTaskStatus) {
    // todo error handling is sketchy, improve
    let mut savepoint = conn.savepoint().unwrap();
    match try_perform_migration(&mut savepoint, migration, builder) {
        Ok(b) => {
            let status = if dry_run {
                savepoint.rollback().unwrap();
                drop(savepoint);
                MigrationTaskStatus::DryRunSuccess
            } else {
                savepoint.commit().unwrap();
                MigrationTaskStatus::Success
            };
            (Ok(b), status)
        }
        Err(e) => {
            let status = if dry_run {
                savepoint.rollback().unwrap();
                drop(savepoint);
                MigrationTaskStatus::DryRunFailure
            } else {
                savepoint.commit().unwrap();
                MigrationTaskStatus::Failure
            };
            (Err(e), status)
        }
    }
}

pub(super) fn update_task_status(
    conn: &rusqlite::Connection,
    job_id: i64,
    status: MigrationTaskStatus,
    error: Option<&str>,
) -> Result<(), Error> {
    conn.execute(
        "UPDATE __libsql_migration_tasks SET status = ?, error = ? WHERE job_id = ?",
        (status as u64, error, job_id),
    )?;

    Ok(())
}

fn try_perform_migration<B: QueryResultBuilder>(
    savepoint: &mut Savepoint,
    migration: &Program,
    builder: B,
) -> Result<B, Error> {
    let mut vm = Vm::new(
        builder, // todo use proper builder
        migration,
        |_| (false, None),
        |_, _, _| (),
    );

    while !vm.finished() {
        vm.step(savepoint).unwrap(); // return migration error
    }

    Ok(vm.into_builder())
}
