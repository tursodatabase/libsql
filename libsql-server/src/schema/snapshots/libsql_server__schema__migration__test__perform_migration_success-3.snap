---
source: libsql-server/src/schema/migration.rs
expression: schema
---
Ok(
    [
        "{Ok(\"type\"): (Text, \"table\"), Ok(\"name\"): (Text, \"__libsql_migration_tasks\"), Ok(\"tbl_name\"): (Text, \"__libsql_migration_tasks\"), Ok(\"rootpage\"): (Integer, 2), Ok(\"sql\"): (Text, \"CREATE TABLE __libsql_migration_tasks (\\n            job_id INTEGER PRIMARY KEY,\\n            status INTEGER,\\n            migration TEXT NOT NULL,\\n            error TEXT\\n        )\")}",
    ],
)
